src/middlewared/middlewared/main.py

```diff
@@ -506,22 +506,25 @@ def do_copy():
             return resp
 
         def copy():
-            while True:
-                read = asyncio.run_coroutine_threadsafe(
-                    filepart.read_chunk(1048576),
-                    loop=self.loop,
-                ).result()
-                if read == b'':
-                    break
-                job.pipes.input.w.write(read)
+            try:
+                try:
+                    while True:
+                        read = asyncio.run_coroutine_threadsafe(
+                            filepart.read_chunk(1048576),
+                            loop=self.loop,
+                        ).result()
+                        if read == b'':
+                            break
+                        job.pipes.input.w.write(read)
+                finally:
+                    job.pipes.input.w.close()
+            except BrokenPipeError:
+                pass
 
         try:
             job = await self.middleware.call(data['method'], *(data.get('params') or []),
                                              pipes=Pipes(input=self.middleware.pipe()))
-            try:
-                await self.middleware.run_in_thread(copy)
-            finally:
-                await self.middleware.run_in_thread(job.pipes.input.w.close)
+            await self.middleware.run_in_thread(copy)
```

Comment:
Should we even wait for this coroutine to finish? Or just let it run in the "background" with ensure_future?

Reply:
Yes, we need, it's a request handler, we can't allow it to finish until request reading is complete.

Downside is that /_upload won't return job ID until job consumes input but for our current use cases it's OK.

Reply:
Derp, obviously. Stupid question :D